### ISSUE TITLE
Can't ban certain types with the ban-types rule

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#2493]
### Bug Report

- __TSLint version__: 4.5.1
- __TypeScript version__: 2.2.1
- __Running TSLint via__: gulp-tslint

#### TypeScript code being linted

```ts
let foo1: {};
let foo2: angular.IPromise<any>;
let foo3: ng.IPromise<any>;
```

with `tslint.json` configuration:

```json
"ban-types": [
    true,
    [ "{}", "Do not use {} as a type since it is not a good definition of what this represents. Please properly type the object instead." ],
    [ "(angular|ng)\\.IPromise<any>", "If this promise does not return anything, use 'angular.IPromise<void>' here instead." ]
],
```
This came about due to issue #2491 that I recently opened.  I had hoped that I could at least lint against these types, but I cannot!  

I would expect violations from the `ban-types` rule to be thrown, but none are.  it seems that the typedef `{}` is not possible to lint against and generics are not possible to be included either.

**Comment by chrismbarr** [comment]
Experimenting with this locally I found that on [line 66 of this rule](https://github.com/palantir/tslint/blob/master/src/rules/banTypesRule.ts#L66) we have
```ts
const typeName = node.typeName.getText();
```
if it were changed to 
```ts
const typeName = node.getText();
```
then it would return the full type including the generics.  This still doen't not report the `{}` type though for whatever reason.

**Comment by ajafff** [comment]
{} is not reported because the implementation looks for TypeReference, but
{} is a TypeLiteral

Am 04.04.2017 16:28 schrieb "Chris Barr" <notifications@github.com>:

Experimenting with this locally I found that on line 66 of this rule
<https://github.com/palantir/tslint/blob/master/src/rules/banTypesRule.ts#L66>
we have

const typeName = node.typeName.getText();

if it were changed to

const typeName = node.getText();

then it would return the full type including the generics. This still
doen't not report the {} type though for whatever reason.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/palantir/tslint/issues/2493#issuecomment-291517244>,
or mute
the thread

**Comment by chrismbarr** [comment]
@ajafff Yeah, i discovered that and updated my comment above.

**Comment by mscharley** [comment]
I'm also noticing I can't catch usages of `any` such as:

```
const foo: any = "hello world!";
```

With the introduction of `unknown` in TS 3.0 I want to deprecate usage of `any` in my project and instantly reached for ban-types.

**Comment by marvinside** [comment]
@mscharley 
Just saw this. There is an extra rule for any, just use
```json
"no-any": true,
```

**Comment by JoshuaKGoldberg** [comment]
@chrismbarr I get the following output from your example:

```powershell
PS C:\Code\test> tslint -c tslint.json -p tsconfig.json

ERROR: C:/Code/test/index.ts[2, 11]: Don't use 'angular.IPromise<any>' as a type. If this promise does not return anything, use 'angular.IPromise<void>' here instead.
ERROR: C:/Code/test/index.ts[3, 11]: Don't use 'ng.IPromise<any>' as a type. If this promise does not return anything, use 'angular.IPromise<void>' here instead.
```

...so it looks like it's just `{}` that's missing complaints, right?

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!
